vespalib/src/vespa/vespalib/net/emulated_epoll.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "emulated_epoll.h"
+#include <chrono>
+#include <cstring>
+#include <vector>
+
+namespace vespalib {
+
+Epoll::Epoll()
+    : _monitorlock(),
+      _wakeup(),
+      _monitored()
+{
+}
+
+Epoll::~Epoll() = default;
+
+uint32_t maybe(uint32_t value, bool yes) { return yes ? value : 0; }
+
+void
+Epoll::add(int fd, void *ctx, bool read, bool write)
+{
+    epoll_event evt;
+    evt.events = maybe(EPOLLIN, read) | maybe(EPOLLOUT, write);
+    evt.data.ptr = ctx;
+    std::lock_guard guard(_monitorlock);
+    _monitored[fd] = evt;
+    _wakeup.write_token();
+}
+
+void
+Epoll::update(int fd, void *ctx, bool read, bool write)
+{
+    epoll_event evt;
+    evt.events = maybe(EPOLLIN, read) | maybe(EPOLLOUT, write);
+    evt.data.ptr = ctx;
+    std::lock_guard guard(_monitorlock);
+    _monitored[fd] = evt;
+    _wakeup.write_token();
+}
+
+void
+Epoll::remove(int fd)
+{
+    epoll_event evt;
+    memset(&evt, 0, sizeof(evt));
+    std::lock_guard guard(_monitorlock);
+    _monitored.erase(fd);
+    _wakeup.write_token();
+}
+
+size_t
+Epoll::wait(epoll_event *events, size_t max_events, int timeout_ms)
+{
+    size_t evidx = 0;
+    std::vector<pollfd> fds;
+    bool allowRetry = true;
+    auto entryTime = std::chrono::steady_clock::now();
+    int timeout_ms_initial = timeout_ms;
```

Comment:
consider keeping the parameter constant and modifying the local variable instead